gcsio/src/main/java/com/google/cloud/hadoop/gcsio/GoogleCloudStorageImpl.java

```diff
@@ -279,137 +276,138 @@ public GoogleCloudStorageImpl(GoogleCloudStorageOptions options, Credentials cre
+   * @param options options to customize behavior
+   * @param httpRequestInitializer request initializer used to initialize all HTTP requests
+   * @param downscopedAccessTokenFn Function that generates downscoped access token
+   * @throws IOException on IO error
    */
-  @SuppressWarnings("CheckReturnValue")
-  @VisibleForTesting
-  public GoogleCloudStorageImpl(GoogleCloudStorageOptions options, Storage storage) {
-    this(options, storage, null, tryGetCredentialsFromStorage(storage));
-    warnIfTracingEnabled(storageOptions.isTraceLogEnabled());
+  public GoogleCloudStorageImpl(
+      GoogleCloudStorageOptions options,
+      HttpRequestInitializer httpRequestInitializer,
+      Function<List<AccessBoundary>, String> downscopedAccessTokenFn)
+      throws IOException {
+    this(
+        options,
+        /* credentials= */ null,
+        /* httpTransport= */ null,
+        httpRequestInitializer,
+        downscopedAccessTokenFn);
   }
 
   /**
    * Constructs an instance of GoogleCloudStorageImpl.
    *
-   * @param options options to customize behavior
+   * @param options {@link GoogleCloudStorageOptions} to use to initialize the object
    * @param credentials OAuth2 credentials that allows access to GCS
-   * @param downscopedAccessTokenFn Function that generates downscoped access token.
+   * @param httpTransport transport used for HTTP requests
+   * @param httpRequestInitializer request initializer used to initialize all HTTP requests
+   * @param downscopedAccessTokenFn Function that generates downscoped access token
    * @throws IOException on IO error
    */
-  public GoogleCloudStorageImpl(
+  @VisibleForTesting
+  GoogleCloudStorageImpl(
       GoogleCloudStorageOptions options,
       Credentials credentials,
+      HttpTransport httpTransport,
+      HttpRequestInitializer httpRequestInitializer,
       Function<List<AccessBoundary>, String> downscopedAccessTokenFn)
       throws IOException {
     logger.atFiner().log("GCS(options: %s)", options);
 
-    this.storageOptions = checkNotNull(options, "options must not be null");
-    this.storageOptions.throwIfNotValid();
-    this.httpRequestInitializer = getHttpRequestInitializers(options, credentials);
+    checkNotNull(options, "options must not be null").throwIfNotValid();
+    // If credentials var is null then use httpRequestInitializer to initialize it
+    if (credentials == null && httpRequestInitializer != null) {
+      checkArgument(
+          httpRequestInitializer instanceof RetryHttpInitializer,
+          "httpRequestInitializer (%s) should be an instance of RetryHttpInitializer for"
+              + " credentials initialization",
+          httpRequestInitializer.getClass());
+      credentials = ((RetryHttpInitializer) httpRequestInitializer).getCredentials();
+    }
+    // If httpRequestInitializer var is null then use credentials to initialize it
+    if (httpRequestInitializer == null) {
+      httpRequestInitializer =
```

Comment:
updating input parameter is a code smell. Especially since there is an instance variable with the same name.

Reply:
Done.